Feat/i2c

DhruvaG2000 opened the pull request.

DhruvaG2000 committed:
Start adding IIC Support

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
feat/I2C: add all function prototypes

* Blank templates of functions necessary for I2c are ready.
* The code compiled so far
* TODO: Start adding actual zephyr implementations

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
README: remove irrelevant line

DhruvaG2000 committed:
minor additions

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
README: Add setup instructions

DhruvaG2000 committed:
README: add docs link

DhruvaG2000 committed:
i2c: mark functions reqd for ADS1115 driver

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Feat/gpio: Add basic GPIO support (#4)

* Add GPIO basic support

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* Add Arduino style blinky example

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* pinMode: add LED_BUILTIN check

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* GPIO: Replace pinNum hardcode to generalise it

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* Supports all GPIOs to be i/p or o/p mode

* Arduino nano ble sense 33 specific
* Tested and works for Digital pins under gpio0 and gpio1 reference in the DT

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* gpio: irrelevant comment deleted

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* feat/gpio: implement digitalRead

* Also fixes other functions to use legacy non-dt functions
This way multiple gpios can be configured and used easily.

This commit was successfully built and tested using a button press LED ON example

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* Fix code formatting

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* gpio: Pulldown input gpios default

And a minor fix to remove build warning

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

* gpio: define LED_BUILTIN in arduino map file

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

alvarowolfx commented on libraries/Wire/Wire.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void arduino::ZephyrI2C::begin() {
+  // const struct device *dev = DEVICE_DT_GET_ANY(adi_adxl372);
+  *i2c_dev = DEVICE_DT_GET(DT_NODELABEL(i2c0));
```
Needs to create the `bus` with the i2c device + device address, like this https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L153
```
struct i2c_dt_spec bus;
```

alvarowolfx commented on libraries/Wire/Wire.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void arduino::ZephyrI2C::setClock(uint32_t freq) {}
+
+void arduino::ZephyrI2C::beginTransmission(uint8_t address) { // TODO for ADS1115
```
Needs to create the `bus` with the i2c device + device address, like this https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L153
```
struct i2c_dt_spec bus;
```

alvarowolfx replied:
sets `_addr` and reset `txBuf`

alvarowolfx commented on libraries/Wire/Wire.cpp:
```diff
@@ -0,0 +1,57 @@
+
+uint8_t arduino::ZephyrI2C::endTransmission(bool stopBit) { return 2; }
+
+uint8_t arduino::ZephyrI2C::endTransmission(void) { // TODO for ADS1115
```
Call  `i2c_write_dt`

```
bus = i2c_dt_spec{ _addr + i2c device }
ret = i2c_write_dt(bus, _txBuf, sizeof(_txBuf));
```

alvarowolfx commented on libraries/Wire/Wire.cpp:
```diff
@@ -0,0 +1,57 @@
+size_t arduino::ZephyrI2C::write(uint8_t data) {  // TODO for ADS1115 
+  return 1; }
+
+size_t arduino::ZephyrI2C::write(const uint8_t *buffer, size_t size) {
```
append data to `txBuf` 

```
txBuf[usedTxBuffer++] = data;
```

alvarowolfx commented on libraries/Wire/Wire.cpp:
```diff
@@ -0,0 +1,57 @@
+size_t arduino::ZephyrI2C::requestFrom(uint8_t address, size_t len,
+                                       bool stopBit) {
+
+  return len;
```
Try to use `i2c_burst_read_dt` to read multiple bytes at the same time
Example: https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L179

alvarowolfx replied:
Zephyr has Ring Buffers too, so we can replicate the MBED implementation 

https://docs.zephyrproject.org/3.0.0/reference/data_structures/ring_buffers.html
MBED: 

* https://github.com/arduino/ArduinoCore-mbed/blob/1da4e665159e7133280c5c33426fb6c22581dc7d/libraries/Wire/Wire.h#L79
* https://github.com/arduino/ArduinoCore-mbed/blob/1da4e665159e7133280c5c33426fb6c22581dc7d/libraries/Wire/Wire.cpp#L119

DhruvaG2000 committed:
gpio: Add button press led example

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Revert "Feat/gpio: Add basic GPIO support (#4)"

This reverts commit 775d8f864b50818cae586655a2795a04957ed7a4.

DhruvaG2000 committed:
Revert "Feat/gpio: Add basic GPIO support (#4)"

This reverts commit 775d8f864b50818cae586655a2795a04957ed7a4.

DhruvaG2000 committed:
i2c: pt1: implement basic wrapper funcs.

* begin()
* beginTransmission()
* Also adds a private section in the class containing few buffer variables

This reverts commit b811145d5775ca0f912d5968d59922dd9729f96e.

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
i2c: implement write_bytes

Whether to use this or just i2c_write directly needs to be thought upon

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
i2c: minor fix

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
i2c: minor fix to write_bytes call

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 commented:
planning to add this example:
```C++
#include <Wire.h>

void setup() {
Serial.begin(115200);
Wire.begin();

Wire.beginTransmission(0x53);
Wire.write(0x2C); 
Wire.write(0x08); 
Wire.endTransmission();

Wire.beginTransmission(0x53);
Wire.write(0x31); 
Wire.write(0x08); 
Wire.endTransmission();

Wire.beginTransmission(0x53);
Wire.write(0x2D); 
Wire.write(0x08); 
Wire.endTransmission();
}

void loop() {
Wire.beginTransmission(0x53);
Wire.write(0x32); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte x0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x33); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte x1 = Wire.read();
x1 = x1 & 0x03;

uint16_t x = (x1 << 8) + x0;
int16_t xf = x;
if(xf > 511)
{
xf = xf - 1024;
}
float xa = xf * 0.004;
Serial.print("X = "); 
Serial.print(xa);
Serial.print(" g"); 
Serial.println(); 


Wire.beginTransmission(0x53);
Wire.write(0x34); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte y0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x35); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte y1 = Wire.read();
y1 = y1 & 0x03;

uint16_t y = (y1 << 8) + y0;
int16_t yf = y;
if(yf > 511)
{
yf = yf - 1024;
}
float ya = yf * 0.004;
Serial.print("Y = "); 
Serial.print(ya);
Serial.print(" g"); 
Serial.println(); 

Wire.beginTransmission(0x53);
Wire.write(0x36); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte z0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x37); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte z1 = Wire.read();
z1 = z1 & 0x03;

uint16_t z = (z1 << 8) + z0;
int16_t zf = z;
if(zf > 511)
{
zf = zf - 1024;
}
float za = zf * 0.004;
Serial.print("Z = "); 
Serial.print(za);
Serial.print(" g"); 
Serial.println(); 
Serial.println(); 
delay(500);
}
```

DhruvaG2000 committed:
i2c: Get it working to a bare minimum stage

* As of now the code was tested and there is I2C activity at the hardware end.
* Also getting responses from the sensor.

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
i2c: Get it to work for just adxl345 single r/w

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Removed Arduino API files

restrucured the entire project to be separate from Arduino Core API code

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Removed Arduino API files

restrucured the entire project to be separate from Arduino Core API code

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Removed Arduino API files

restrucured the entire project to be separate from Arduino Core API code

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
Readme: Update installation instructions

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

beriberikix committed:
Update license to Apache 2.0

All LGPL code removed and all code is original, Apache 2.0 code.

DhruvaG2000 committed:
update zephyr/module.yml to include examples

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
update README: west.yml instructions

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 committed:
i2c: implement ring buffer

Also an effort to try and support multiple byte r/w (unntested)

Tested and working on ADXL345

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

DhruvaG2000 commented:
Added a much cleaner PR here: https://github.com/zephyrproject-rtos/gsoc-2022-arduino-core/pull/8

Closing this because this repo will not be used in near future.
